### PR TITLE
Dynamic GitHub username fetch

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

@mxstbr this lets users refresh their github username from the client, as well as uses the redis cache to server the username based on an initial network request to the GitHub api.

### CODE REVIEW

**Comment by brianlovin** [comment]
@mxstbr could you help me out with those flow errors in the test_static - it looks like ioredis' typedefs are saying we're doing caching wrong, but we haven't had this problem with the same code in `api/queries/community/metaData` (for example) - maybe we need to actually update our `cache.set()` and `cache.get()` args? Fwiw this code works locally with my local redis server

File: api/authentication.js [hunk 1]
```diff
@@ -304,6 +309,19 @@ const init = () => {
...
           // if the user already has a githubProviderId, don't override it
           if (req.user.githubProviderId) {
+            /*
+              Update the cached content of the github profile that we store
+              in redis for the graphql resolver. This allows us to put a button
+              on the client for a user to re-connect a github profile from
+              the web app which will update the cache with any changed usernames
+            */
+            await cache.set(
+              `githubProfile:${req.user.id}`,
```
Comment (mxstbr) [inline api/authentication.js]: nit: let's key this as `user:${req.user.id}:github`

EDIT: We should probably move this into a shared util in `shared/` so the key is consistent between all invocations even if we change it in the future.

File: api/queries/user/githubProfile.js [hunk 2]
```diff
@@ -1,10 +1,50 @@
...
-    id: githubProviderId,
-    username: githubUsername,
-  };
+export default async ({ githubProviderId, id }: DBUser) => {
+  if (!githubProviderId) return null;
+
+  const cachedGithubProfile = await cache.get(`githubProfile:${id}`);
+
+  if (cachedGithubProfile) {
+    const parsed = JSON.parse(cachedGithubProfile);
```
Comment (mxstbr) [inline api/queries/user/githubProfile.js]: nit: `try/catch` this JSON.parse and, if it errors, go fetch it again to make sure bad data in the cache doesn't hurt users.

Probably something like:

```JS
const cachedGitHubProfile = ...

if (cachedGitHubProfile) {
  try {
    const parsed = JSON.parsed(cachedGitHubProfile);
    return { ... };
  // Just ignore errors, causing the `return` not to run and skipping below the if block
  } catch(err) {}
}

// ...other code here
const apiUrl = ...;
```

File: api/queries/user/githubProfile.js [hunk 3]
```diff
@@ -1,10 +1,50 @@
...
+    return {
+      id: parsed.id,
+      username: parsed.username,
+    };
+  } else {
+    const apiUrl = 'https://api.github.com';
+    const apiRoute = '/user/';
+    const { GITHUB_OAUTH_CLIENT_ID, GITHUB_OAUTH_CLIENT_SECRET } = process.env;
+    const queryString = `?client_id=${GITHUB_OAUTH_CLIENT_ID}&client_secret=${GITHUB_OAUTH_CLIENT_SECRET}`;
```
Comment (mxstbr) [inline api/queries/user/githubProfile.js]: nit: let's move the static variables outside of the resolver to the top of the file

File: api/queries/user/githubProfile.js [hunk 4]
```diff
@@ -1,10 +1,50 @@
...
+    const fetchUrl = apiUrl + apiRoute + githubProviderId + queryString;
+
+    const networkGithubProfile = await fetch(fetchUrl)
+      .then(res => res.json())
+      .catch(err => {
+        console.error('Error getting github info from github api', err);
+        return null;
+      });
+
+    if (networkGithubProfile) {
```
Comment (mxstbr) [inline api/queries/user/githubProfile.js]: nit: let's invert this `if` to make the nesting more readable

```JS
if (!networkGitHubProfile) return null;

// ...other code
```

File: api/queries/user/githubProfile.js [hunk 5]
```diff
@@ -1,10 +1,50 @@
...
+      const cacheRecord = {
+        id: networkGithubProfile.id,
+        username: networkGithubProfile.login,
+      };
+
+      await cache.set(
+        `githubProfile:${id}`,
```
Comment (mxstbr) [inline api/queries/user/githubProfile.js]: nit: need to update this cache key too

File: api/queries/user/githubProfile.js [hunk 6]
```diff
@@ -1,10 +1,50 @@
...
-export default ({ githubProviderId, githubUsername }: DBUser) => {
-  if (!githubProviderId || !githubUsername) return null;
-  return {
-    id: githubProviderId,
-    username: githubUsername,
-  };
+export default async ({ githubProviderId, id }: DBUser) => {
+  if (!githubProviderId) return null;
+
+  const cachedGithubProfile = await cache.get(`githubProfile:${id}`);
```
Comment (mxstbr) [inline api/queries/user/githubProfile.js]: nit: this cache key also needs to be updated

**Review by mxstbr — COMMENTED** [review]
Code looks great apart from a couple nitpicks! :clap:

Why shouldn't we store the username in the db and then add the button to the frontend to update to fresh GitHub data? That would allow people who change their GitHub username to refresh the username, while we don't have to incur all the complexity of caching it in Redis.

As far as I can tell, the benefit of caching is that it updates automatically after the cache expires, but I personally don't think that's worth the extra complexity of caching. Open to disagreement on that though!

**Comment by mxstbr** [comment]
I _think_ the flow errors might go away if you use `"EX"`  instead of `"ex"`?

File: api/authentication.js (hunk 1 above)
Comment (brianlovin) [reply api/authentication.js]: Will do!

**Comment by brianlovin** [comment]
> As far as I can tell, the benefit of caching is that it updates automatically after the cache expires, but I personally don't think that's worth the extra complexity of caching. Open to disagreement on that though!

That's the main reason: if people update their GitHub username, it will be resolved automatically on Spectrum within a day. Since most people might not think to try and update their GitHub username on every 3rd party platform they've connected to, having this be automatic could be nice.

That being said: it does add some extra complexity with caching logic, which doesn't feel *too* bad to me right now, but if you think it's overdoing it I suppose we could store in the db. I don't have a strong feeling on this.

**Comment by mxstbr** [comment]
I don't think it's worth the extra complexity to be honest?

**Comment by brianlovin** [comment]
Sure, will fix accordingly :)

**Comment by brianlovin** [comment]
Okay, last few commits remove all the caching - now we just let people refresh from the client and in the authentication code we check for a different incoming username :)

File: api/queries/channel/metaData.js [hunk 7]
```diff
@@ -28,7 +27,13 @@ export default async (root: DBChannel, _: any, ctx: GraphQLContext) => {
 
   // Cache the fields for an hour
   (await typeof cachedOnlineMemberCount) === 'number' ||
-    cache.set(`channel:${id}:onlineMemberCount`, onlineMemberCount, 'ex', 3600);
+    cache.set(
+      channelOnlineMemberCount(id),
+      onlineMemberCount,
+      'NX',
```
Comment (mxstbr) [inline api/queries/channel/metaData.js]: I don't think we want NX here, why did you add that? I don't know what that does.

File: api/queries/community/metaData.js [hunk 8]
```diff
@@ -36,12 +40,13 @@ export default async (root: DBCommunity, _: any, ctx: GraphQLContext) => {
   // Cache the fields for an hour
   await Promise.all([
     typeof cachedChannelCount === 'number' ||
-      cache.set(`community:${id}:channelCount`, channelCount, 'ex', 3600),
+      cache.set(communityChannelCount(id), channelCount, 'NX', 'EX', 3600),
```
Comment (mxstbr) [inline api/queries/community/metaData.js:43]: Same here, what's NX?

File: api/queries/channel/metaData.js (hunk 7 above)
Comment (brianlovin) [reply api/queries/channel/metaData.js]: `NX` is part of the redis signature now. `NX` says: if the record for this key doesn't exist, create it. `EX` following says 'set the expieration for this key'

Comment (brianlovin) [reply api/queries/channel/metaData.js]: https://stackoverflow.com/a/16296998

**Review by mxstbr — DISMISSED** [review]
Nice, LGTM man! :100:

File: api/queries/channel/metaData.js (hunk 7 above)
Comment (brianlovin) [reply api/queries/channel/metaData.js]: Actually I think that comment is misleading...

**Comment by brianlovin** [comment]
So I'm looking at the `flow-typed` defs for `ioredis` and it's very confusing. `.set` has this signature:
```js
set(
      key: string,
      value: string,
      nx?: "NX",
      ex?: "EX",
      expires?: number,
      callback?: ResCallbackT<string>
    )
```

This is why if we don't have `NX` it throws a flow error, because it's expecting an `NX` argument in the position that we're putting `EX` - this is dumb, because these are optional args (see: https://github.com/luin/ioredis/issues/159#issuecomment-238015453) 

It seems that the solution is to remove `NX` (we actually *don't* want this behavior*) and `$FlowIssue` the resulting errors of putting `EX` as the third argument...

**Comment by brianlovin** [comment]
I'm upgrading the `ioredis` package to the latest (I don't see any breaking changes that impact us in the changelog). However, they don't have a valid flow type definitions file for our version of flow (`0.66.0`). So I'm going to `create-stub` this.

**Comment by brianlovin** [comment]
Let me know how all these changes look to you @mxstbr - no rush. Can ship this later this week.

**Review by mxstbr — DISMISSED** [review]
LGTM
